### ISSUE TITLE
Failing test:  gocql panics in E2E test suite with `panic: scylla: <ip>:9042 invalid number of shards`

### ISSUE THREAD

**Issue by rzetelskik** [scylladb/scylla-operator#1399]
We've consistently been hitting a panic from gocql in our E2E test suite on master.

https://github.com/scylladb/scylla-operator/actions/runs/6178751661/job/16772628746#step:3:3702
```
E0914 00:01:49.220811       1 tests/tests_run.go:447] "Process failed" err="process failed" ID=6 Command="/usr/bin/scylla-operator-tests run scylla-operator/conformance/parallel --loglevel=2 --artifacts-dir=/home/runner/work/_temp/e2e-artifacts --flake-attempts=5 --timeout=110m --feature-gates= --override-ingress-address=10.101.34.231 --parallelism=8 --parallel-server-address=127.0.0.1:34643 --loglevel=0 --random-seed=1694646811 --parallel-shard=6" ProcessState="exit status 2" Logs=<
...
	panic: scylla: 10.105.211.70:9042 invalid number of shards
	
	goroutine 997 [running]:
	github.com/gocql/gocql.(*scyllaConnPicker).Put(0xc0005b1440, 0xc000904000)
		github.com/gocql/gocql@v1.6.0/scylla.go:400 +0x448
	github.com/gocql/gocql.(*hostConnPool).connect(0xc00058b1f0)
		github.com/gocql/gocql@v1.6.0/connectionpool.go:554 +0x2b0
	github.com/gocql/gocql.(*hostConnPool).fill(0xc00058b1f0)
		github.com/gocql/gocql@v1.6.0/connectionpool.go:397 +0x159
	github.com/gocql/gocql.(*policyConnPool).addHost(0xc0003decd0, 0xc000b849c0)
		github.com/gocql/gocql@v1.6.0/connectionpool.go:255 +0x1cf
	github.com/gocql/gocql.(*Session).startPoolFill(0xc000685000, 0x0?)
		github.com/gocql/gocql@v1.6.0/events.go:214 +0x2d
	created by github.com/gocql/gocql.(*controlConn).setupConn
		github.com/gocql/gocql@v1.6.0/control.go:301 +0x26a
 >
Error: all processes have finished but the suite still isn't done
```

https://github.com/scylladb/scylla-operator/suites/16143937950/artifacts/921837495

See also:
https://github.com/scylladb/scylla-operator/actions/runs/6191772438/job/16810747030#step:3:3178
https://github.com/scylladb/scylla-operator/actions/runs/6203601128/job/16844545678#step:3:3315

**Comment by rzetelskik** [comment]
Unfortunately I haven't been able to backtrack it to a single test so far - the logs printed out by a failed test are only the common messages printed by helpers, so they don't contain the ginkgo node's logs and are pretty much useless. I've tried dry running the suite with the same seed and parallelism to at least see the subset of specs running in this process but it doesn't seem to be constant in my dry runs for some reason.

I think the easiest modification we could do to at least help us backtrack it to a single test is to modify our framework to annotate the namespaces with the parallel process ordinal as well as the full text of the ginkgo leaf; it's very easy to do and should help narrowing it down. @tnozicka wdyt?

**Comment by tnozicka** [comment]
sounds like a good start

**Comment by rzetelskik** [comment]
Hit it again https://github.com/scylladb/scylla-operator/actions/runs/6228943103/job/16906764197#step:3:3253.
```
E0918 23:59:44.878396       1 tests/tests_run.go:447] "Process failed" err="process failed" ID=1 Command="/usr/bin/scylla-operator-tests run scylla-operator/conformance/parallel --loglevel=2 --artifacts-dir=/home/runner/work/_temp/e2e-artifacts --flake-attempts=5 --timeout=110m --feature-gates= --override-ingress-address=10.96.205.75 --parallelism=8 --parallel-server-address=127.0.0.1:33695 --loglevel=0 --random-seed=1695079112 --parallel-shard=1" ProcessState="exit status 2" Logs=<
        ...
	Sep 18 23:22:38.117: INFO: Created namespace "e2e-test-scyllacluster-r8dxc-8856q".
	...
	Sep 18 23:23:43.454: INFO: Reading data from table "29x29vx8"."test"
	2023/09/18 23:23:43 gocql: unable to dial control conn 10.107.122.144:9042: dial tcp :0->10.107.122.144:9042: connect: connection refused
	2023/09/18 23:23:43 gocql: unable to connect to any ring node: dial tcp :0->10.107.122.144:9042: connect: connection refused
	2023/09/18 23:23:43 gocql: control falling back to initial contact points.
        ...
	panic: scylla: 10.107.122.144:9042 invalid number of shards
	
	goroutine 474 [running]:
	github.com/gocql/gocql.(*scyllaConnPicker).Put(0xc0007a5170, 0xc000823520)
		github.com/gocql/gocql@v1.6.0/scylla.go:400 +0x448
	github.com/gocql/gocql.(*hostConnPool).connect(0xc00043d1f0)
		github.com/gocql/gocql@v1.6.0/connectionpool.go:554 +0x2b0
	github.com/gocql/gocql.(*hostConnPool).fill(0xc00043d1f0)
		github.com/gocql/gocql@v1.6.0/connectionpool.go:397 +0x159
	github.com/gocql/gocql.(*policyConnPool).addHost(0xc0001f6eb0, 0xc000902340)
		github.com/gocql/gocql@v1.6.0/connectionpool.go:255 +0x1cf
	github.com/gocql/gocql.(*Session).startPoolFill(0xc000861800, 0x0?)
		github.com/gocql/gocql@v1.6.0/events.go:214 +0x2d
	created by github.com/gocql/gocql.(*controlConn).setupConn
		github.com/gocql/gocql@v1.6.0/control.go:301 +0x26a
 >
 ```
 
 ```
 $ yq -e '.items[] | select(.metadata.name=="e2e-test-scyllacluster-r8dxc-8856q")' e2e-artifacts/cluster-scoped-resources/namespaces.yaml
 apiVersion: v1
kind: Namespace
metadata:
  annotations:
    ginkgo-full-text: ScyllaCluster should reconcile resource changes
    ginkgo-parallel-process: "1"
  creationTimestamp: "2023-09-18T23:22:38Z"
  labels:
    e2e: scylla-operator
    framework: scyllacluster-r8dxc
    kubernetes.io/metadata.name: e2e-test-scyllacluster-r8dxc-8856q
  name: e2e-test-scyllacluster-r8dxc-8856q
  resourceVersion: "3777"
  uid: 4b47f81e-99e6-4192-96ef-803b6b6a0eff
spec:
  finalizers:
    - kubernetes
status:
  phase: Active
 ```

**Comment by rzetelskik** [comment]
The issue reproduces locally in `ScyllaCluster should reconcile resource changes` E2E test.

**Comment by rzetelskik** [comment]
Bisecting leads to the gocql bump introducing a regression - `first bad commit: [4625fdc1b0a6df89d23d83c66e0803a113dbb0d1] Merge pull request #1363 from tnozicka/bump-deps`.

Running `6b310ee0` (last good commit on master) with just gocql bumped to v1.11.1 reproduces the issue.

**Comment by mykaul** [comment]
@avelanarius - FYI

**Comment by rzetelskik** [comment]
@mykaul @avelanarius I'll create an issue in gocql repository, atm I'm trying to get some more context - e.g. why it hasn't been panicking in presubmits ran with a different CI stack.

**Comment by tnozicka** [comment]
@rzetelskik kacper good job tracing this all the way. For this repo the flake was fixed in #1408 so I'll close this one as done. The actual issue is tracked in https://github.com/scylladb/gocql/issues/145 and I've created https://github.com/scylladb/scylla-operator/issues/1415 for us to bump afterwards.
